[TexMap] Implement damage information scaling along with viewport

Scony opened the pull request:
#### 5b7de9ac7f51d4802dd241bc86ce3613c11bb8bf
<pre>
[TexMap] Implement damage information scaling along with viewport
<a href="https://bugs.webkit.org/show_bug.cgi?id=285601">https://bugs.webkit.org/show_bug.cgi?id=285601</a>

Reviewed by Carlos Garcia Campos.

This change:
- fixes the damage of layers on transform changes
- fixes the damage so that it uses correct transforms
- fixes the damage of layers on size changes
- fixes the damage of layers with m_backingStoreProxy or m_contentsBuffer that have no changes recorded
thus effectively adding a support for damage information scaling along with viewport.

* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::computeTransformsRecursive):
(WebCore::TextureMapperLayer::prepareForPainting):
(WebCore::TextureMapperLayer::paint):
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::damageWholeLayerDueToTransformChange):
(WebCore::TextureMapperLayer::transformRectForDamage):
(WebCore::TextureMapperLayer::setSize):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::canInferDamage const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::flushCompositingState):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):

Canonical link: <a href="https://commits.webkit.org/288924@main">https://commits.webkit.org/288924@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0eb53e0d2b6df626fa5ac5ed24aeb96f5a5baba7

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/84196 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/3817 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/38500 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/89272 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/35205 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/86281 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/3905 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/11790 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/65491 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/23329 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/87242 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/2930 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/76510 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/45784 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/2878 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/30740 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/34254 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/73795 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/31504 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/90653 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/11463 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/8322 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/73944 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/11689 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/72335 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/73147 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/18203 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/17458 "Passed tests") | [⏳ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/macOS-Ventura-Release-WK2-Intel-Tests-EWS "Waiting to run tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/2826 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/11415 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/16891 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/11264 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/14740 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/13037 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
```diff
@@ -113,6 +114,7 @@ class TextureMapperLayer : public CanMakeWeakPtr<TextureMapperLayer> {
     bool syncAnimations(MonotonicTime);
     WEBCORE_EXPORT bool descendantsOrSelfHaveRunningAnimations() const;
 
+    WEBCORE_EXPORT void prepareForPainting(TextureMapper&);
```
I don't understand why you have to change the public API. After this change, all client have to call `prepareForPainting`. Why not `paint` calls `prepareForPainting` internally?

Scony replied:
Good point, I can call `prepareForPainting` in `paint` iff `!ENABLE(DAMAGE_TRACKING)`

fujii replied:
I don't understand why it's only for `!ENABLE(DAMAGE_TRACKING)`. 
Does `beginClip` have to be called between `beginPainting` and `prepareForPainting` if ENABLE(DAMAGE_TRACKING)?

Scony replied:
Well, in case of `ENABLE(DAMAGE_TRACKING)` the mandatory sequences are:

1. `beginPainting` -> `beginClip` -> `collectDamage`
2. `prepareForPainting` -> `collectDamage`
3. `prepareForPainting` -> `paint`

In the above:

- `beginPainting` and `beginClip` set the clip region to the latest viewport size; therefore when `collectDamage` is invoked, it can trim damage rectangles to the clip region.
- `prepareForPainting` calculates transforms of all layers recursively; therefore when `collectDamage` is invoked, it can position damage rectangles correctly (in the viewport coordinate space)

In case of `!ENABLE(DAMAGE_TRACKING)` everything can be done as it was before.

fujii replied:
Got it.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -341,14 +348,20 @@ void TextureMapperLayer::computeTransformsRecursive(ComputeTransformData& data)
 
+void TextureMapperLayer::paint(TextureMapper& textureMapper)
+{
+#if !ENABLE(DAMAGE_TRACKING)
+    prepareForPainting(textureMapper);
+#endif
```
Damage tracking is also a runtime decision, shouldn't we do this only when using the damage? I think it's clearer if we handle this in the caller, because in CoordinatedGraphicsScene::paintToCurrentGLContext it's not obvious that is is called when not tracking damage.

Scony replied:
This is actually just a split of `paint()` into `prepareForPainting()` and `paint()` for the purpose of `ENABLE(DAMAGE_TRACKING)` and technically it could be done unconditionally (even outside of `ENABLE(DAMAGE_TRACKING)`). However, as @fujii mentioned, it's better to have `prepareForPainting()` as part of `paint()` for non-damage usecases as it's safer for callers of `paint()`.

Morever, I did an unconditional split in previous iteration, but windows builds started failing then, so it looks like windows builds rely on having `prepareForPainting()` within `paint()`.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -413,16 +429,33 @@ void TextureMapperLayer::collectDamageSelf(TextureMapperPaintOptions& options, D
+{
+    // When the layer's transform changes, we must not only damage whole layer using new transform,
+    // but also using old transform to cover the area not affected by layer anymore.
+    if (beforeChange != afterChange && !m_damage.isInvalid()) {
```
Do an early return here, or move this check to the caller and remove the IfNeeded since this is only caled in one place. I would also check if damage is invalid first, before comparing the matrix.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
```diff
@@ -1185,6 +1218,13 @@ void TextureMapperLayer::setBoundsOrigin(const FloatPoint& boundsOrigin)
+    if (m_state.size != size) {
+        // When layer size changes, we damage whole layer for now.
+        // FIXME: Damage only affected area.
+        m_damage.add(FloatRect(FloatPoint::zero(), size));
```
Can m_damage be invalid here? Maybe we should do this only when tracking damaging?

Scony replied:
Good point.
Regarding doing this only when damage propagation is on - it will require coordination of implementation with the damage flow from `GraphicsLayerCoordinated` so I propose to do it in a separate change.

carlosgcampos commented on Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
```diff
@@ -932,6 +932,14 @@ void CoordinatedPlatformLayer::flushCompositingState()
+        // FIXME: Remove special handling of content layers.
+        layer.setDamage({ });
+    }
+#endif
```
Can we handle this before when we set the damage? The contents layer hasn't been set yet, but we know if the layer will have it 

```
if (m_committedContentsBuffer || (m_imageBackingStore && m_imageBackingStoreVisible))
```

carlosgcampos commented on Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
```diff
@@ -51,9 +51,12 @@ void CoordinatedGraphicsScene::paintToCurrentGLContext(const TransformationMatri
+
+    std::optional<FloatRoundedRect> rectContainingRegionThatActuallyChanged;
 #if ENABLE(DAMAGE_TRACKING)
+    currentRootLayer.prepareForPainting(*m_textureMapper);
```
Don't we need to move this inside the if?

Scony replied:
assuming the split of `paint()` into `prepareForPainting()` and `paint()` is done only for `ENABLE(DAMAGE_TRACKING)` - it must stay within guards.

Scony committed:
[TexMap] Implement damage information scaling along with viewport
https://bugs.webkit.org/show_bug.cgi?id=285601

Reviewed by Carlos Garcia Campos.

This change:
- fixes the damage of layers on transform changes
- fixes the damage so that it uses correct transforms
- fixes the damage of layers on size changes
- fixes the damage of layers with m_backingStoreProxy or m_contentsBuffer that have no changes recorded
thus effectively adding a support for damage information scaling along with viewport.

* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.cpp:
(WebCore::TextureMapperLayer::computeTransformsRecursive):
(WebCore::TextureMapperLayer::prepareForPainting):
(WebCore::TextureMapperLayer::paint):
(WebCore::TextureMapperLayer::collectDamageSelf):
(WebCore::TextureMapperLayer::damageWholeLayerDueToTransformChange):
(WebCore::TextureMapperLayer::transformRectForDamage):
(WebCore::TextureMapperLayer::setSize):
* Source/WebCore/platform/graphics/texmap/TextureMapperLayer.h:
(WebCore::TextureMapperLayer::canInferDamage const):
* Source/WebCore/platform/graphics/texmap/coordinated/CoordinatedPlatformLayer.cpp:
(WebCore::CoordinatedPlatformLayer::flushCompositingState):
* Source/WebKit/WebProcess/WebPage/CoordinatedGraphics/CoordinatedGraphicsScene.cpp:
(WebKit::CoordinatedGraphicsScene::paintToCurrentGLContext):

Canonical link: https://commits.webkit.org/288924@main
